scripts/gen_idt.py

```diff
@@ -0,0 +1,250 @@
+#!/usr/bin/env python3
+#
+# Copyright (c) 2017 Intel Corporation
+#
+# SPDX-License-Identifier: Apache-2.0
+
+import argparse
+import sys
+import struct
+import os
+from elftools.elf.elffile import ELFFile
+from elftools.elf.sections import SymbolTableSection
+
+# This will never change, first selector in the GDT after the null selector
+KERNEL_CODE_SEG = 0x08
+
+# These exception vectors push an error code onto the stack.
+ERR_CODE_VECTORS = [8, 10, 11, 12, 13, 14, 17]
+
+def debug(text):
+    if not args.verbose:
+        return
+    sys.stdout.write(os.path.basename(sys.argv[0]) + ": " + text + "\n")
+
+def error(text):
+    sys.stderr.write(os.path.basename(sys.argv[0]) + ": " + text + "\n")
+    sys.exit(1)
+
+# See Section 6.11 of the Intel Architecture Software Developer's Manual
+gate_desc_format = "<HHBBH"
+
+def create_irq_gate(handler, dpl):
+    present = 1
+    gate_type = 0xE # 32-bit interrupt gate
+    type_attr = gate_type | (dpl << 5) | (present << 7)
+
+    offset_hi = handler >> 16
+    offset_lo = handler & 0xFFFF
+
+    data = struct.pack(gate_desc_format, offset_lo, KERNEL_CODE_SEG, 0,
+            type_attr, offset_hi)
+    return data
+
+
+def create_task_gate(tss, dpl):
+    present = 1
+    gate_type = 0x5 # 32-bit task gate
+    type_attr = gate_type | (dpl << 5) | (present << 7)
+
+    data = struct.pack(gate_desc_format, 0, tss, 0, type_attr, 0)
+    return data
+
+
+def create_idt_binary(idt_config, filename):
+    with open(filename, "wb") as fp:
+        for handler, tss, dpl in idt_config:
+            if handler and tss:
+                error("entry specifies both handler function and tss")
+
+            if not handler and not tss:
+                error("entry does not specify either handler or tss")
+
+            if (handler):
```

Comment:
Style nitpick: unneeded parenthesis. There are other instances of this in this file.